File: Hash.java
class Hash {
    public  int tableSize;
    public  int overflowSize;
    private int realsize;
    private int posOverflow; // control insertions on overflow area
    public  Object htable[]; // make easy to substitute the type of the table

    public Hash() {
        this(10, 10);
    }

    public Hash(int tableSize, int overflowSize) {
        this.tableSize = tableSize;
        this.overflowSize = overflowSize;
        this.realsize = tableSize + overflowSize;
        this.posOverflow = 0;
        this.htable = new Object[realsize];
        for (int i = 0; i < realsize; i++) 
            htable[i] = null;
    }

    public int hash(Object item) {
        // change the key with whatever you want
        return (item.key % tableSize);
    }

    // return the position where item has been inserted
    public int insert(Object item) {
        int pInsertion = hash(item); 
        int posReturn = -1; 
        // if the position is free
        if (htable[pInsertion] == null) {
            htable[pInsertion] = item.clone();
            posReturn = pInsertion;
        } else {
            if (posOverflow < tableSize) {
                htable[tableSize + posOverflow] = item.clone();
                posReturn = tableSize + posOverflow;
                posOverflow++;
            } else {
                System.out.print("Área de reserva cheia! Falha ao inserir ");
                item.imprimir();
            }
        }
        return posReturn;
    }

    // return the position of the removed element
    // -1 if not found
    public int remove(Object toBeRemoved) {
        int pos = -1;
        int pRemotion = hash(toBeRemoved);
        // found in the main area
        if (htable[pRemotion].equals(toBeRemoved)) {
            pos = pRemotion;
            boolean hasSubstitute = false;
            // change the item in the main area
            // with the item in the overflow area
            // with the same hash 
            for (int i = 0; i < posOverflow && !hasSubstitute; i++) {
                if (hash(htable[i + tableSize]) == pRemotion) {
                    hasSubstitute = true;
                    htable[pRemotion] = htable[i + tableSize].clone();
                    if (posOverflow > 0) 
                        posOverflow--;
                    htable[i + tableSize] = htable[tableSize + posOverflow].clone();
                    htable[tableSize + posOverflow] = null;
                }
            }
        } else {
            // not found in main area -> search in the overflow area
            for (int i = 0; i < posOverflow; i++) {
                if (htable[i + tableSize].equals(toBeRemoved)) {
                    pos = i + tableSize;
                    if (posOverflow > 0)
                        posOverflow--;
                    htable[i + tableSize] = htable[posOverflow].clone();
                    htable[posOverflow] = null;
                    i = posOverflow;
                }
            }
        }
        return pos;
    }

    // return the removed item
    // null if not found
    public Object remove(int key) {
        if (key < 0 || key >= realsize) {
            System.out.println("Chave de remocao invalida.");
            return null;
        }

        Object removed = null;
        if (htable[key] != null) {
            removed = htable[key].clone();
            boolean hasSubstitute = false;
            // again, put in the main area
            // the item with the same hash coming
            // from the overflow area
            for (int i = 0; i < posOverflow && !hasSubstitute; i++) {
                if (hash(htable[i + tableSize]) == key) {
                    hasSubstitute = true;
                    htable[key] = htable[i + tableSize].clone();
                    if (posOverflow > 0) 
                        posOverflow--;
                    htable[i + tableSize] = htable[posOverflow].clone();
                    htable[posOverflow] = null;
                }
            }
        }
        return removed;
    }

    public void showItens() {
        System.out.println("\nÁrea principal: [");
        for (int i = 0; i < tableSize; i++) {
            System.out.print("\t");
            if (htable[i] != null)
                System.out.println(htable[i].toString());
            else
                System.out.println("null");
        }
        System.out.println("]\n");
        System.out.println("Área de reserva: [");
        for (int i = 0; i < posOverflow; i++) {
            System.out.print("\t"+htable[tableSize + i].toString() + "\n");
        }
        System.out.println("]\n");
    }

    public boolean search(Object obj) {                                                        
        boolean retorno = false;                                                              
        if (obj == null) {                                                                      
            System.out.println("Object nulo!");                                              
        } else {                                                                              
            int hashJ = hash(obj);                                                              
            if (htable[hashJ].equals(obj)) {                                                    
                retorno = true;                                                               
            } else {                                                                          
                // searching on overflow area                                                 
                for (int i = 0; i < posOverflow && !retorno; i++) {                           
                    if (htable[i + tableSize].equals(obj)) {                                    
                        retorno = true;                                                       
                        }                                                                         
                }                                                                                                                                                         
            }                                                                                 
        }
        return retorno;                                                                       
    }

    // return the position of the item
    // -1 if not found
    public int searchWithPos(Object obj) {
        int pos = -1;
        if (j == null) {
            System.out.println("Object nulo!");
        } else {
            int hashJ = hash(obj);
            if (htable[hashJ].equals(obj)) {
                pos = hashJ;
            } else {
                // searching in overflow area
                for (int i = 0; i < posOverflow && (pos == -1); i++) {
                    if (htable[i + tableSize].equals(obj)) {
                        pos = i + tableSize;
                    }
                }
            }
        }
        return pos;
    }
}

File: hash.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#define TAM 7
#define REALSIZE (TAM+5)

int reserva = 0;

int ** makeTable() {
    int **thash = (int **) malloc(sizeof(int *) * REALSIZE);
    for (int i = 0; i < REALSIZE; i++) {
        thash[i] = NULL;
    }
    return thash;
}

int hash(int key) {
    return key % TAM;
}

// return the position where the item
// has been inserted
// return -1 if could not insert
int inserir(int **thash, int item) {
    int posInsercao = -1;
    int pos = hash(item);
    int *p = (int *) malloc(sizeof(int));
    *p = item;
    if (thash[pos] == NULL) {
        thash[pos] = p;
        posInsercao = pos;
    } else {
        if (reserva < REALSIZE) {
            int posReserva = TAM + reserva++;
            thash[posReserva] = p; 
            posInsercao = posReserva;
        } else {
            printf("Area reserva cheia!\n");
            free(p);
        }
    }
    p = NULL;
    return posInsercao;
}

// remove item using its position
int removerPos(int **thash, int key) {
    int removed = -9999;

    // is key valid?
    if (key < 0 || key >= REALSIZE) {
        printf("Chave de procura invalida.\n");
    } else {
        // if there is something in the position
        if (thash[key] != NULL) {
            removed = *thash[key];
            bool hasSubstitute = false;
            for (int i = 0; i < reserva && !hasSubstitute; i++) {
                if (hash(*thash[i + TAM]) == key) {
                    hasSubstitute = true;
                    *thash[key] = *thash[i + TAM];
                    if (reserva > 0)
                        reserva--;
                    *thash[i + TAM] = *thash[TAM + reserva];
                    free(thash[TAM + reserva]);
                    thash[TAM + reserva] = NULL;
                }
            }
            if (!hasSubstitute) {
                free(thash[key]);
                thash[key] = NULL;
            }
        }
    }

    return removed;
}

// remove item -> return its position 
int removerItem(int **thash, int item) {
    int posRemovido = -1;
    int posItem = hash(item);
    if (*thash[posItem] == item) {
        posRemovido = posItem;
        bool hasSubstitute = false;
        for (int i = 0; i < reserva && !hasSubstitute; i++) {
            // if any item has the same hash
            // as the item to be removed
            if (hash(*thash[i + TAM]) == posItem) {
                hasSubstitute = true;
                *thash[posItem] = *thash[i + TAM];
                if (reserva > 0)
                    reserva--;
                *thash[i + TAM] = *thash[TAM + reserva];
                free(thash[TAM + reserva]);
                thash[TAM + reserva] = NULL;
            }
        }
        if (!hasSubstitute) {
            free(thash[posItem]);
            thash[posItem] = NULL;
        }
    } else {
        // searching in the overflow area
        for (int i = 0; i < reserva; i++) {
            if (*thash[i + TAM] == item) {
                posRemovido = i + TAM;
                // if reserva > 0, the position of
                // the last item in overflow area
                // is TAM + reserva - 1
                if (reserva > 0)
                    reserva--;
                *thash[i + TAM] = *thash[TAM + reserva];
                free(thash[TAM + reserva]);
                thash[TAM + reserva] = NULL;
                i = reserva;
            }
        }
    }
    return posRemovido;
}

void mostrar(int **thash) {
    printf("Área principal: [ ");
    for (int i = 0; i < TAM; i++) 
        printf("%d ", (thash[i] == NULL) ? -1 : *thash[i]);
    printf("]\n");
    
    printf("Área de reserva: [ ");
    for (int i = 0; i < (REALSIZE - TAM); i++) 
        printf("%d ", (thash[i + TAM] == NULL) ? -1 : *thash[i + TAM]);
    printf("]\n\n");
}

int main() {
    int **thash = makeTable();
    inserir(thash, 10);
    inserir(thash, 30);
    mostrar(thash);
    inserir(thash, 20);
    inserir(thash, 21);
    inserir(thash, 14);
    inserir(thash, 45);
    inserir(thash, 35);
    inserir(thash, 4);
    inserir(thash, 12);
    inserir(thash, 1);
    mostrar(thash);
    printf("Posicao removido: %d\n", removerItem(thash, 21));
    printf("apos remocao do 21: \n");
    mostrar(thash);
    printf("Posicao removido: %d\n", removerItem(thash, 45));
    printf("apos remocao do 45: \n");
    printf("Item removido: %d\n", removerPos(thash, 30));
    mostrar(thash);
}
